RecoEcal/EgammaClusterProducers/src/EGRefinedSCFixer.cc

```diff
@@ -0,0 +1,491 @@
+    auto& newConv(fixedConvs->back());
+    // conversion keeps a PtrVector, but according to PFEGammaProducer, single-leg conversions
+    // always have only one SC ptr.
+    reco::CaloClusterPtrVector scPtrVec;
+    // we create the new refined SC collection in the same order as the old one
+    scPtrVec.push_back(reco::CaloClusterPtr(scHandle, conv.caloCluster()[0].key()));
+    newConv.setMatchingSuperCluster(scPtrVec);
+  }
+
+  iEvent.put(std::move(fixedConvs));
+}
+
+
+std::vector<edm::Ptr<reco::CaloCluster> >
+EGRefinedSCFixer::getSubClustersMissing(const reco::SuperCluster& lhs,
+					const reco::SuperCluster& rhs)
+{
+  std::vector<edm::Ptr<reco::CaloCluster> > missingSubClusters;
+  for(auto& subClus : lhs.clusters()){
+    auto compFunc=[&subClus](const auto& rhs){return subClus->seed()==rhs->seed();};
+    if(std::find_if(rhs.clusters().begin(),rhs.clusters().end(),compFunc)==rhs.clusters().end()){ 
+      missingSubClusters.push_back(subClus);
+    }
+  }
+  return missingSubClusters;
+}
+
+std::unordered_set<unsigned>
+EGRefinedSCFixer::getListOfClusterSeedIdsForNewSC(const reco::SuperCluster& orgRefinedSC,
+						  const reco::SuperCluster& orgSC,
+						  const reco::SuperCluster& fixedSC,
+						  const std::vector<edm::Ptr<reco::CaloCluster> >& clustersAddedToAnySC)
+  
+{
+  auto clusAdded = getSubClustersMissing(orgRefinedSC,orgSC);
+  auto clusRemoved = getSubClustersMissing(orgSC,orgRefinedSC);
+
+  
+  std::unordered_set<unsigned> detIdsOfClustersForNewSC;
+  for(auto& clus : fixedSC.clusters()){
+    auto compFunc=[&clus](auto& rhs){return rhs->seed().rawId()==clus->seed().rawId();};
+    
+    //first check if the cluster was removed by the refining process
+    bool notRemoved = std::find_if(clusRemoved.begin(),clusRemoved.end(),compFunc)==clusRemoved.end();
+
+    //now check if it was assigned to another supercluster (it wont be picked up by the previous
+    //check if the parent supercluster picked it up during reclustering)
+    //if its a cluster which is added to any supercluster it wont add it
+    //its okay if it was added to this supercluster as we will add it in below
+    bool notAddedToASuperCluster = std::find_if(clustersAddedToAnySC.begin(),
+						clustersAddedToAnySC.end(),compFunc)==clustersAddedToAnySC.end();
+ 
+    if(notRemoved && notAddedToASuperCluster){
+      detIdsOfClustersForNewSC.insert(clus->seed().rawId());
+    }
+  }
+  for(auto clus : clusAdded){
+    detIdsOfClustersForNewSC.insert(clus->seed().rawId());
+  }
+  
+  return detIdsOfClustersForNewSC;
+}
+
+std::vector<edm::Ptr<reco::PFCluster> >
+EGRefinedSCFixer::getClustersFromSeedIds(const std::unordered_set<unsigned>& seedIds,const edm::Handle<edm::View<reco::PFCluster> >& inClusters)
+{
+  std::vector<edm::Ptr<reco::PFCluster> > outClusters;
+  for(size_t clusNr=0;clusNr<inClusters->size();clusNr++){
+    edm::Ptr<reco::PFCluster> clusPtr(inClusters,clusNr);
+    if(seedIds.count(clusPtr->seed().rawId())>0){
```

Comment:
the Ptr can be made inside the "if" to reduce unnecessary creation and dereferencing